DS-55 [가게] Service 클래스 설계 및 정의

y00njinuk opened the pull request:
## Goals
- 가게 서비스 클래스를 설계 및 정의한다.

## Implements
- [x] 가게 서비스 클래스 정의 및 구현
- [x] 일부 네이밍 변경이 필요한 필드 수정 

## Related Issue 
- https://dangol-sonnim.atlassian.net/browse/DS-55?atlOrigin=eyJpIjoiMmM3NGEyOGFjMDNiNDQxMTliZDc0NzI5ZTZmMTFmMjYiLCJwIjoiaiJ9

## Test
- "새로운 가게 정보를 전달받으면 가게를 생성하도록 한다."
- "가게 ID 값을 전달받으면 가게를 정상적으로 조회할 수 있다"
- "수정할 가게 정보를 전달받으면 가게를 수정하도록 한다."

y00njinuk committed:
refactor: rename field of Store Entity and add Access level

[issue-key] DS-55
- 일부 필드에 대한 네이밍 수정
- Setter 접근 권한을 막기 위해 AccessLevel 어노테이션 추가

y00njinuk committed:
feat: define DTO classes of Store

[issue-key] DS-45
- 가게와 관련된 DTO 클래스 정의

y00njinuk committed:
feat: add repoistory interface of Store

[issue-key] DS-55
- 가게와 관련된 신규 레포지토리 계층 인터페이스 추가

y00njinuk committed:
feat: add service of Store

[issue-key] DS-55
- 가게와 관련된 신규 서비스 계층 인터페이스 추가

y00njinuk committed:
feat: add ErrorCodeMessage "STORE_NOT_FOUND"

[issue-key] DS-55
- 존재하지 않는 가게에 대한 에러메시지 명세 추가

y00njinuk committed:
test: add unit test of Store service

[issue-key] DS-55
- 가게 서비스 테스트 추가

rnjstmdals6 commented on src/main/java/com/dangol/dangolsonnimbackend/store/service/impl/StoreServiceImpl.java:
```diff
@@ -0,0 +1,89 @@
+
+        if(store.isPresent()) {
+            store.get().setName(dto.getName());
+            store.get().setPhoneNumber(dto.getPhoneNumber());
```
1. 해당 dto 값이 null 일 경우 엔티티 필드에 null 이 설정되지않나요?
2. 트랜잭션 처리를 하지 않으신 이유가 있으실까요?

y00njinuk replied:
1. 그렇군요. DTO 객체에 값이 Null일 경우 엔티티의 not null 정책을 만족하지 못하네요. 다시 수정해보도록 하겠습니다.
2. Transcational 어노테이션을 사용하란 말씀이신건가요? 롤백을 위해서?

rnjstmdals6 replied:
네 롤백을 위해선데 정확히는 작업의 완정성을 보장하기 위함입니다!
업데이트 서비스에서 일부 작업이 된 상태에 예외가 발생된다면
작업 전체가 실패해야 되는 것이 맞지 않나 생각합니다.

y00njinuk replied:
넵 추가하도록 하겠습니다.

rnjstmdals6 commented on src/test/java/com/dangol/dangolsonnimbackend/store/service/StoreServiceTest.java:
```diff
@@ -0,0 +1,96 @@
+    public void givenStoreId_whenFindById_thenReturnStore() {
+        StoreResponseDTO response = storeService.signup(dto);
+
+        assertTrue(storeQueryRepository.findById(response.getId()).isPresent());
```
해당 부분은 레포지토리 테스트에 가깝지 않을까요!?

y00njinuk replied:
테스트의 의도는 요청(StoreSignupRequestDTO)으로 들어온 값과 응답으로 받은 값(StoreResponseDTO)을 비교하고 이들의 데이터가 서로 동일하다는 것을 나타내기 위함이였는데 흠.. 다른 좋은 방법이 있을까요? 중간에 Repository를 거치지 않으면 이 둘의 값이 서로 같다는 걸 나타낼 수 있는 방법이 떠오르질 않더라구요.

rnjstmdals6 replied:
개인적인 생각으로는 서비스에 있는 매서드를 사용해서 엔티티를 불러와 비교하는 것이 서비스 테스트에 더 적합할 것 같아요
서비스에 있는 FindById 의 반환타입을 Entity로 수정하면 조금 더 수월하게 테스트 가능 할 것 같아요

y00njinuk replied:
넵. 서비스 쪽에서 엔티티 객체를 다루는 것이 일반적인 거 같네요. 수정하도록 하겠습니다.

rnjstmdals6 commented on src/main/java/com/dangol/dangolsonnimbackend/store/service/impl/StoreServiceImpl.java:
```diff
@@ -0,0 +1,89 @@
+        Optional<Store> store = storeQueryRepository.findById(id);
+
+        return store.map(StoreResponseDTO::new)
+                .orElseThrow(() -> new BadRequestException(ErrorCodeMessage.STORE_NOT_FOUND));
```
가게를 ResponseDTO로 매핑하는 과정 중 예외가 던져졌을 때 NOT_FOUND 메세지를 던지는 게 적절 할 지 궁금합니다.

y00njinuk replied:
서버 측의 오류로 인해서 발생한 것인데 NOT FOUND 의미가 조금 안맞긴 하네요. 다른 걸로 다시 검토해보겠습니다. 피드백 감사합니다!

rnjstmdals6 replied:
가게를 조회하는 부분에도 404처리를 해주는 것이 졸아보입니다!

rnjstmdals6 commented on src/main/java/com/dangol/dangolsonnimbackend/store/domain/Store.java:
```diff
@@ -59,14 +58,15 @@ public class Store {
     private Long categoryId;
 
     @Column(nullable = false, unique = true)
+    @Setter(AccessLevel.NONE)
```
AccessLevel을 NONE으로 사용하신 이유에 대해 설명 들을 수 있을까요?

y00njinuk replied:
특정 컬럼에 대해서는 Setter에 대한 접근권한을 막기 위해 사용해보았습니다.

rnjstmdals6 commented on src/main/java/com/dangol/dangolsonnimbackend/store/domain/Store.java:
```diff
@@ -2,14 +2,12 @@
 
 @Entity
-@Getter
+@Data
```
@Data 어노테이션을 사용하신 이유가 궁금합니다.

y00njinuk replied:
@Data 어노테이션이 @Getter와 @Setter 어노테이션을 포함한다고 하여 해당 어노테이션을 사용하였습니다.

rnjstmdals6 replied:
@Setter 의 사용 용도가 궁금합니다!
Setter를 사용하게 되면 의도하지 않은 변경으로 인해 오류를 발생시킬 수 있고
객체지향 특성인 캡슐화를 지키기 위해서 Entity에서는 Setter 사용을 최대한 지양하고 있습니다.

값을 변경할 목적이 있는 경우에만 따로 매서드를 만들어 변경가능하게 하면 Setter 없이도 충분히 구현가능하지 않을까 싶습니다!

마지막으로 @Data 어노테이션은 Entity에서 사용하면 안되는 어노테이션으로 알고 있습니다.
데이터 어노테이션은 게터 세터 뿐만 아니라 @ToString, @EqualsAndHashCode 와 같은 어노테이션도 포함합니다.
@ToString 은 필요하지 않은 문자열 연산을 진행하므로 성능 상 안좋은 영향을 끼칩니다.
또한  @EqualsAndHashCode은 객체간 동일성을 비교하는 어노테이션입니다. 엔티티에서는 양방향 매핑을 하게 될 가능성이 많습니다. 양방향 관계일 때 사용하게 되면 무한 순환 참조 문제가 발생할 수 있다고 합니다.

y00njinuk replied:
결국, public한 setter 메소드와 불필요한 어노테이션으로 인해 여러 문제가 발생하는군요. getter/setter 정의를 제외하고는 롬복에 대한 편의성이 없어지는 느낌이 드네요.

rnjstmdals6 replied:
맞습니다 롬복 사용에는 상당한 주의가 필요하다고 합니다. 관련해서 한번 찾아보시는 걸 추천드릴게요!

rnjstmdals6 commented on src/main/java/com/dangol/dangolsonnimbackend/errors/enumeration/ErrorCodeMessage.java:
```diff
@@ -9,7 +9,9 @@ public enum ErrorCodeMessage {
+    STORE_NOT_FOUND("존재하지 않는 가게입니다.");
 
-    private String message;
+    private final String message;
```
👍👍👍

y00njinuk committed:
feat: add ErrorCodeMessage "RESPONSE_CREATE_ERROR"

[issue-key] DS-55
- 가게 서비스 레이어에서 신규 에러 메세지 정의

y00njinuk committed:
feat: add exception for handling internal server error (500)

[issue-key] DS-55
- 500 상태 코드에 대한 에러 처리를 위해서 예외 클래스 정의

y00njinuk committed:
refactor: convert to use optional field

[issue-key] DS-55
- 필수적으로 값을 세팅해야 하는 필드를 제외하고는 모두 Optional 타입으로 변경

y00njinuk committed:
refactor: add convert to Store entity from StoreUpdateDTO

[issue-key] DS-55
- DTO 값으로부터 Optional한 엔티티 클래스로 전환 추가
- null 체크에 대한 검사 추가

y00njinuk committed:
refactor: fix to make StoreResponseDTO

[issue-key] DS-55
- StoreResponseDTO 응답 객체를 생성하는 방식 수정
- 임의의 값에만 할당하고 나머지 필드에 대해서는 null 체크를 통해 필터링이 모두 가능하도록 변경함을 확인

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/errors/enumeration/ErrorCodeMessage.java:
```diff
@@ -6,12 +6,17 @@
+    STORE_NOT_FOUND("존재하지 않는 가게입니다."),
+
+    // 500
+    RESPONSE_CREATE_ERROR("응답 데이터를 생성하던 도중 에러가 발생하였습니다.");
```
응답데이터를 생성하면서 발생한 오류이기에 위와 같은 에러 메세지로 정의하였습니다.

rnjstmdals6 replied:
500번대도 추가하셨군요 ㅎㅎ 좋습니다!

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/errors/RestApiExceptionHandler.java:
```diff
@@ -21,6 +21,16 @@ public ResponseEntity<ErrorResponse> handleBadRequestException(BadRequestExcepti
+        log.info("Internal Server Error, requestId={}, message={}", requestId, ex.getMessage(), ex);
+
+        ErrorResponse errorResponse = new ErrorResponse(ex.getMessage(), requestId);
+        return new ResponseEntity<>(errorResponse, HttpStatus.INTERNAL_SERVER_ERROR);
```
InternalServerException 예외 발생 시에는 서버에서 상태코드는 500으로 반환합니다.

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/store/domain/Store.java:
```diff
@@ -83,4 +85,21 @@ public Store(StoreSignupRequestDTO dto) {
         this.name = name;
     }
+
+    public Optional<Store> update(StoreUpdateDTO dto) {
```
Store 객체 수정 시 Optional 필드에 대한 null 체크를 모두 확인하고 수정하도록 합니다.

rnjstmdals6 replied:
Optional로 null 체킹이 조금 더 세련되어졌네요!

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/store/service/impl/StoreServiceImpl.java:
```diff
@@ -67,24 +68,12 @@ public StoreResponseDTO findById(Long id) {
-        }
-
-        return store.map(StoreResponseDTO::new)
+        return storeQueryRepository.findByRegisterNumber(dto.getRegisterNumber())
```
registerNumber 값은 필수적으로 가져야 하기에 객체에 대한 null 체크와 registerNumber에 대한 null 체크를 진행 후에 수정 작업을 진행하고 응답 데이터를 생성하도록 수정하였습니다.

rnjstmdals6 replied:
404예외를 추가하는 건 어떨까요?

y00njinuk replied:
혹시 이유가 있을까요~? 서버 측에서 응답 데이터를 생성하다가 발생한 오류이기 때문에 500으로 정의하는게 어떨까? 라고 생각했거든요.

rnjstmdals6 replied:
스트림 처리 전 findByRegisterNumber 요청했을 때 DB에 요청한 데이터가 없다면 404의 에러를 던져줘야 하지 않을까 싶습니다!
이후 데이터 처리 시 발생한 오류에 대해서는 500이 적절한 것 같습니다

y00njinuk replied:
아하, 그러면 분기 처리를 해두도록 하겠습니다.

rnjstmdals6 commented on src/main/java/com/dangol/dangolsonnimbackend/store/domain/Store.java:
```diff
@@ -83,4 +85,21 @@ public Store(StoreSignupRequestDTO dto) {
+        dto.getCategoryId().ifPresent(categoryId -> this.categoryId = categoryId);
+        dto.getRegisterName().ifPresent(registerName -> this.registerName = registerName);
+
+        return Optional.of(this);
```
Optional로 반환하는 부분은 불필요하지 않을까요?
void로 반환해도 필드 수정은 이루어질 텐데 
void로 반환하지 않은 이유가 궁금합니다.

y00njinuk replied:
chaining 연산을 위해서 Optional로 모두 통일하였습니다.

rnjstmdals6 commented on src/main/java/com/dangol/dangolsonnimbackend/store/dto/StoreUpdateDTO.java:
```diff
@@ -0,0 +1,57 @@
+        this.registerNumber = registerNumber;
+    }
+
+    public StoreUpdateDTO name(String name) { this.name = Optional.of(name); return this; }
```
Optional로 사용하는 것도 좋지만.. 이렇게 UpdateDTO를 작성하는 경우 에너지가 많이 소비되고
생산성이 떨어지지 않을까요?

y00njinuk replied:
맞아요. 저도 작성하는 동안 힘이 들었긴 했는데 테스트를 위한 코드로 구현을 하였거든요. 또한 객체 생성 시 빌더 패턴을 활용하기 위한 목적으로 추가해둔 것도 있습니다. 서비스 테스트 시에만 필요한 코드이긴 하겠네요.

rnjstmdals6 replied:
아하 그렇군요 알겠습니다!

rnjstmdals6 commented on src/main/java/com/dangol/dangolsonnimbackend/store/service/impl/StoreServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+     */
+    @Override
+    public StoreResponseDTO findById(Long id) {
+        Optional<Store> store = storeQueryRepository.findById(id);
```
null 일 경우에 404 에러를 던져주면 어떨까요?

y00njinuk replied:
🆗

y00njinuk committed:
refactor: check for the result of QueryDSL is not empty

[issue-key] DS-55
- QueryDSL 결과가 존재하지 않으면 BadRequestException을 발생시키도록 추가

y00njinuk commented on src/main/java/com/dangol/dangolsonnimbackend/store/service/impl/StoreServiceImpl.java:
```diff
@@ -68,8 +71,12 @@ public StoreResponseDTO findById(Long id) {
     public StoreResponseDTO updateStoreByDto(StoreUpdateDTO dto) {
+        Optional<Store> store =  storeQueryRepository.findByRegisterNumber(dto.getRegisterNumber());
+
+        if(!store.isPresent())
```
QueryDSL의 결과에 따라서 throw할 exception 유형을 결정합니다.

y00njinuk committed:
Merge branch 'main' into feature/DS-55-add-store-service
